File: bots/intgrah/v54.2.4/builder/state/update/prune.py
from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from cambc import Controller

    from builder.state import State


def prune_stale(state: State, ct: Controller) -> None:
    w = state.w
    nearby_positions = ct.get_nearby_tiles()
    state.nearby_positions = nearby_positions
    state.nearby_buildings = []

    state.healable_buildings = [
        p for p in state.healable_buildings if not ct.is_in_vision(p)
    ]
    state.adjacent_to_enemy_launcher = {
        p for p in state.adjacent_to_enemy_launcher if not ct.is_in_vision(p)
    }
    state.enemy_turret_ray_tiles = {
        p for p in state.enemy_turret_ray_tiles if not ct.is_in_vision(p)
    }
    state.friendly_turret_ray_tiles = {
        p for p in state.friendly_turret_ray_tiles if not ct.is_in_vision(p)
    }

    for pos in nearby_positions:
        if 0 <= pos.x < state.w and 0 <= pos.y < state.h:
            i = pos.y * w + pos.x
            state.conveyors_to_here[i] = [
                p for p in state.conveyors_to_here[i] if not ct.is_in_vision(p)
            ]
            state.splitters_to_here[i] = [
                p for p in state.splitters_to_here[i] if not ct.is_in_vision(p)
            ]

File: bots/rs/drewfett/v1001/src/builder/update/prune.rs
use cambc::Controller;

use crate::builder::Builder;
use crate::util::posint::idx_of;

pub fn prune_stale(builder: &mut Builder, ct: &mut Controller<'_>) {
    builder.nearby_buildings = pyrust::vec::new!();
    builder.visible_ti_ore = pyrust::vec::new!();
    builder.visible_ax_ore = pyrust::vec::new!();
    builder.visible_harvesters = pyrust::vec::new!();

    let mask = pyrust::clone!(builder.vision_mask);
    builder.healable_buildings = pyrust::collect!(pyrust::filter!(
        pyrust::copied!(pyrust::iter!(builder.healable_buildings)),
        |p| mask[idx_of(*p) as usize] == 0
    ));
    let prev_launcher = pyrust::set::clone!(builder.adjacent_to_enemy_launcher);
    pyrust::set::clear!(builder.adjacent_to_enemy_launcher);
    for &p in pyrust::iter!(&prev_launcher) {
        if mask[p as usize] == 0 {
            pyrust::set::add!(builder.adjacent_to_enemy_launcher, p);
        }
    }
    let prev_enemy = pyrust::set::clone!(builder.enemy_turret_ray_tiles);
    pyrust::set::clear!(builder.enemy_turret_ray_tiles);
    for &p in pyrust::iter!(&prev_enemy) {
        if mask[p as usize] == 0 {
            pyrust::set::add!(builder.enemy_turret_ray_tiles, p);
        }
    }
    let prev_friendly = pyrust::set::clone!(builder.friendly_turret_ray_tiles);
    pyrust::set::clear!(builder.friendly_turret_ray_tiles);
    for &p in pyrust::iter!(&prev_friendly) {
        if mask[p as usize] == 0 {
            pyrust::set::add!(builder.friendly_turret_ray_tiles, p);
        }
    }
}
